src/main/java/Duke.java

```diff
@@ -1,10 +1,140 @@
+import duke.command.Parser;
+import duke.command.DukeException;
+import duke.command.Storage;
+import duke.command.TaskList;
+import duke.command.Ui;
+import duke.task.Task;
+import duke.task.ToDo;
+import duke.task.Deadline;
+import duke.task.Event;
+
+import java.util.Scanner;
+import java.io.IOException;
+
+/**
+ * Represents an application that that manages <code>ToDo</code>s, <code>duke.task.Event</code>s, and <code>Deadline</code>s.
+ * A Duke object can add <code>Task</code>s, delete them, mark them as Done, and maintain a history of Tasks entered
+ * during earlier execution.
+ */
 public class Duke {
+    private Ui ui;
+    private TaskList tasks;
+    private Storage storage;
+
+    /**
+     * Constructs a Duke object that saves Tasks in the provided filePath.
+     * @param filePath the path of the text file which is a directory for the Tasks.
+     */
+    public Duke(String filePath) {
+        ui = new Ui();
+        storage = new Storage(filePath);
+        try {
+            tasks = new TaskList(storage.load());
+        } catch (IOException de) {
+            ui.showLoadingError();
+            tasks = new TaskList();
+        } catch (DukeException de) {
+            ui.showException(de);
+        }
+    }
+
+    /**
+     * Integrates <code>Parser</code>, <code>TaskList</code>, <code>Storage</code>, and <code>Ui</code> to
+     * deal with User's commands and accordingly manage database.
+     */
+    public void run() {
+        Scanner sc = new Scanner(System.in);
+        ui.showWelcome();
+        while (true) {
+            String command = sc.nextLine();
+            ui.printLine();
+            try {
+                Parser commandAnalyzer = new Parser(command);
+                if (commandAnalyzer.isValid()) {
+                    if (commandAnalyzer.getType().equals("done")) {
+                        int index = Integer.parseInt(commandAnalyzer.getList().get(0)) - 1;
+                        try {
+                            if (index >= tasks.size() || index < 0) {
+                                throw new DukeException(" :( OOPS!!! Requested task number is not available");
+                            }
+                            Task temp = tasks.get(index);
+                            temp.markAsDone();
+                            ui.showTaskDone(temp);
+                            storage.update(tasks);
+                        } catch (DukeException | IOException de) {
+                            ui.showException(de);
+                        } catch (NumberFormatException nfe) {
+                            ui.showNumberFormatError("done");
+                        }
+                    } else if (command.equals("bye")) {
+                        ui.showGoodBye();
+                        break;
+                    } else if (command.equals("list")) {
+                        ui.showTasks(tasks);
```

Comment:
Sorry, I don't understand why is `ui.showTasks(tasks)` called twice 😅

Reply:
Oh shoot! Thanks for finding the typo!